### ISSUE TITLE
Flux getting stuck in case of any java.lang.Error(OOM)

### ISSUE THREAD

**Issue by kamaljit-1991** [reactor/reactor-core#2521]
In parallel flux whenever there is OOM in one of the threads, the whole process is getting stuck until we kill it manually. I have created an example to simulate it. Please refer the following code:

```
@Test
public void test_Flux() {
    Flux.range(1, 30)
            .parallel(6)
            .runOn(Schedulers.fromExecutor(Executors.newFixedThreadPool(6)))
            .doOnNext(this::writeValues)
            .sequential()
            .blockLast();
}

private void writeValues(int val) {
    if (val == 10) {
        throw new OutOfMemoryError("failed here");
        // throw new RuntimeException("failed here");
    }
    System.out.println("the thread name is :" + Thread.currentThread().getName() + " val is :" + val);
    sleep(3);
}

private void sleep(int i) {
    try {
        Thread.sleep(i * 1000L);
    } catch (InterruptedException e) {
        Thread.currentThread().interrupt();
        throw new RuntimeException(e);
    }
}
```
Can we crash here instead of getting stuck?

**Comment by simonbasle** [comment]
`OutOfMemoryError` is considered an unrecoverable exception by Reactor, which tends to give up a throw it. Here it probably ends up in the uncaughtExceptionHandler of the threads from the thread pool, rather than being turned into an `onError` signal that can reach the `blockLast()`, like for the majority of exceptions.

**Comment by kamaljit-1991** [comment]
@simonbasle Then do you suggest to implement `setDefaultUncaughtExceptionHandler()` method for Thread and then do `System.exit()` inside it. Because I don't see any other way here. 

If there is any VM error in any thread at least the process needs to exit instead of getting hanged there. Or is there any better way to do it ??

**Comment by kushagraThapar** [comment]
@simonbasle  - is the only way to handle this is to use `uncaughtExceptionHandler` ?
Because I tried every other way of mapping the error, and the hang just stays on, which is blocking our customers to make any progress if they run into this. 

Please let me know if you would like me to open a new issue for this one.
Here is my sample code to repro this hang. 


```
public static void main(String[] args) throws InterruptedException {

        Flux<Integer> integerFlux = Flux.range(0, 7).map(number -> {
            logger.info("Number is : {}", number);
            if (number > 5) {
                throw new OutOfMemoryError("Custom GC Failure");
            }
            return number;
        }).doOnError(ex -> {
            logger.error("Completed exceptionally", ex);
        }).doOnNext(next -> {
            logger.info("Next is : {}", next);
        }).doOnComplete(() -> {
            logger.info("Completed successfully");
        }).doFinally(signalType -> {
            logger.info("Finally signal is : {}", signalType);
        }).onErrorMap(throwable -> {
            logger.info("On error map", throwable);
            return throwable;
        }).onErrorContinue((throwable, object) -> {
            logger.error("on error continue : {}", object, throwable);
        }).onErrorStop().onErrorReturn(6).onErrorResume(throwable -> {
            logger.info("on error resume", throwable);
            return Mono.error(throwable);
        }).subscribeOn(Schedulers.boundedElastic());

        Iterator<Integer> integers = integerFlux.toIterable().iterator();
        while(integers.hasNext()) {
            logger.info("Next value is : {}", integers.next());
        }

        logger.info("Going to sleep now");

        Thread.sleep(5000);

        logger.info("I woke up");
}
```
